proposals/extension-interfaces.md

```diff
@@ -0,0 +1,194 @@
+2.	If the value of `s` is `null`, `p` should also be `null` inside M2. Moreover, a null check inside `M2` should properly check for the nullability of the original value. If this is not true, the author of `M2` cannot reliably check for null before invoking instance members and the method might cause crashing behavior on valid input.
+3.	`isinst` type checks and casts should be equivalent to the same check on the original value, if the original value directly implemented the interface. Many libraries rely on type checks for correctness and performance – it is not acceptable to invalidate these checks.
+4.	Reflection operations like the `.GetType()` method should return the same results as if the original value directly implemented the interface. Like casts, many libraries rely on reflection access for correct implementation.
+5.	If the original value is a value type, like `int`, a copy operation should be performed. This is consistent with the existing conversion rules for value types, which performs a boxing operation. Boxing produces an observable effect on mutating interface members. Extension interface should not alter that effect.
+
+The above rules state changes to semantics that should not happen due to an extension interface implementation. This appears to be broadly true for the feature. It appears that the only two desired semantic changes are that a conversion should exist an extension is valid, and that the interface implementation in the extension should be invoked through interface dispatch when a direct interface implementation is not otherwise available.
+
+[ ] Open question: Which implementation should be chosen if a direct implementation and an extension implementation are available? Should an error be produced instead?
+
+### Constrained generic parameter
+This case is almost identical to _Interface parameter_, but with a generic type parameter as the parameter type instead.
+```
+string M1(string s, int i) => M2(s) + M2(i);
+string M2<T>(T p) where T : IPrettyPrint { ... }
+```
+Unlike in the Interface parameter example, no conversion is performed. However, in the above example, both `string` and `int` should be considered valid substitutions for the type parameter `T`, even though they would not have originally satisfied the `IPrettyPrint` constraint. Once again, the interface members on `IPrettyPrint` should be invocable on `p` and they should use the implementation provided in the extension implementations.
+The main differences in this example and Interface parameter are around boxing. While a value type would have been copied and boxed with an `IPrettyPrint` parameter type, a generic substitution would normally not box the parameter during the invocation of `M2` or during a call to `PrettyPrint` if the input argument directly implemented `IPrettyPrint`. The same should be true of an extension implementation, and side-effects of `M2` and `PrettyPrint` should be propagated back to the `i` parameter in the call `M2(i)`.
```

Comment:
> The same should be true of an extension implementation, and side-effects of `M2` and `PrettyPrint` should be propagated back to the `i` parameter in the call `M2(i)`.

I don't understand this.
In today's C#, a value type is copied in a method call, even in a generic one.
What do you mean by: should be propagated back to the `i` parameter.

Reply:
On a constrained instance invocation on a generic type parameter, the receiver is a `ref` variable, so the interface invocation is side-effect-preserving.

Reply:
> a constrained instance invocation on a generic type parameter

Can you give an example please, I'm still not sure to fully understand this line :)

Reply:
Yup, so if you have a call like

```C#
interface IFace {
    void Mutate();
}
struct S : IFace {
  public int F;
  public void Mutate() {
    F++;
  }
}
void M<T>(T t) where T : IFace {
  Console.WriteLine(t.F);
  t.Mutate();
  Console.WriteLine(t.F);
}
```

You can see that the call to the interface Method `Mutate` does not copy the struct `t`. Instead, the function mutates the struct variable directly. If you look underneath, you'll see the reason why is that the `this` parameter in the `S.Mutate` method is actually `ref S this`, that is it's passing the struct by-ref.